### PR TITLE
Use .mlx syntax

### PR DESCRIPTION
**Pull request by andreypopp**
**DO NOT MERGE**

This showcases using Melange and Reason React with .mlx syntax (including editor support, though you may need additional configuration for your editor to associate .mlx extension with ocamllsp).

### CODE REVIEW

File: melange-opam-template.opam [hunk 1]
```diff
@@ -7,14 +7,25 @@ license: "MIT"
...
   "ocaml-lsp-server" {dev}
   "dot-merlin-reader" {dev}
   "odoc" {with-doc}
 ]
+pin-depends: [
+  ["dune.3.10.0"                  "git+https://github.com/andreypopp/dune.git#main"]
+  ["melange.1.0.0"                "git+https://github.com/melange-re/melange.git#main"]
+  ["mlx.~dev"                     "https://github.com/andreypopp/mlx/archive/refs/heads/main.zip"]
+  ["ocamlmerlin-mlx.~dev"         "https://github.com/andreypopp/mlx/archive/refs/heads/main.zip"]
+  ["reason.3.8.2"                 "git+https://github.com/reasonml/reason.git#master"]
```
Comment (jchavarri) [inline melange-opam-template.opam:27]: Reason still needed to build reason-react right? but using mlx with melange would not require it?

File: src/ReactApp.mlx [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module App = struct
+  let make () =
+    ["Hello " ^ World.name ^ "!"; "This is React with OCaml and .mlx syntax!"]
+    |> List.map (fun greeting -> <h1> (React.string greeting) </h1>)
+    |> Array.of_list
+    |> React.array [@@react.component]
```
Comment (jchavarri) [inline src/ReactApp.mlx:6]: Pretty sure you had this in mind already, but would be nice to have an API similar to the one jsoo-react [implemented](https://github.com/ml-in-barcelona/jsoo-react/blob/aa0c590469cf3bec0b148615b967bd0903586b3d/test/test_ml.ml#L159):

```ml
let%component make ~a ~b =
  React.Hooks.use_effect ~on:(a, b) (fun () -> incr count);
  div [||] []
```

**Review by jchavarri — COMMENTED** [review]
This is amazing @andreypopp! 😮 

> though you may need additional configuration for your editor to associate .mlx extension with ocamllsp

lol it works perfectly in vscode just by picking "OCaml" on the bottom right "Language Mode", I think the only thing required besides recognizing `mlx` from lsp server is to tweak the syntax highlighting, but otherwise looks pretty much perfect:

mlx

File: melange-opam-template.opam (hunk 1 above)
Comment (andreypopp) [reply melange-opam-template.opam:27]: Yeah, needed for reason-react.

I'm actually wondering if some core libs in Melange ecosystem should be in Reason or pure OCaml. Seems like bindings to DOM and other Web APIs don't require JSX and thus can be in .ml (just less things to build to use them).

File: src/ReactApp.mlx (hunk 2 above)
Comment (andreypopp) [reply src/ReactApp.mlx:6]: Yeah, I have this implemented in [react_of_ocaml](https://github.com/andreypopp/react_of_ocaml/blob/data/example/browser/example.ml#L10). For regular components, async components (for RSC) and for components exported as client components to server (in this case type annotations are required for arguments).

Or are you suggesting to implement this in reason-react ppx as well?

File: melange-opam-template.opam (hunk 1 above)
Comment (jchavarri) [reply melange-opam-template.opam:27]: All Melange "core" libs like `melange.dom` and `melange.node` have always been [written in `ml`](https://github.com/melange-re/melange/tree/b334d3a4d8fbda1b3c4e10939525cc58268544dd/jscomp/others). And some libs like [melange-jest](https://github.com/melange-community/melange-jest/blob/main/jest/jest.ml) are also in ml. But as soon as there was some JSX involved, it was natural to switch to Reason.

I agree that now, with the appearance of `mlx`, there is a much stronger case for writing things like ReactJS bindings in OCaml syntax.

File: src/ReactApp.mlx (hunk 2 above)
Comment (jchavarri) [reply src/ReactApp.mlx:6]: No, this is what I meant. To be honest, I am not sure what's the place for reason-react anymore, after `mlx`. 😕

**Pull request closed without merging**